### PR TITLE
Updates for Android release builds

### PR DESCRIPTION
**Pull request by atavism**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -569,18 +571,28 @@ $(LANTERN_MOBILE_ANDROID_SDK): $(ANDROID_SDK)
...
+android-debug: android-dependencies
+	gradle -PlanternVersion=$(GIT_REVISION) -b $(LANTERN_MOBILE_DIR)/app/build.gradle \
 		clean \
 		assembleDebug
 
-android: $(LANTERN_MOBILE)
+android-release: require-version require-secrets android-dependencies
+	@echo "Generating distribution package for android..."
+	cp $$SECRETS_DIR/android/keystore.release.jks $(LANTERN_MOBILE_DIR)/app
+	gradle -PlanternVersion=$$VERSION -b $(LANTERN_MOBILE_DIR)/app/build.gradle \
```
Comment (atavism) [inline Makefile]: Here we just pass to the gradle the library version number which is accessible via ``BuildConfig.LANTERN_VERSION``

File: src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/model/UI.java [hunk 2]
```diff
@@ -154,16 +152,8 @@ public UI(LanternMainActivity activity, SharedPreferences mPrefs) {
...
-            @Override
-            public void run() {
-                mPrefs.edit().putString("versionNum", latestVersion).commit();
-
-                versionNum.setText(latestVersion);
-            }
-        });
+    public void setVersionNum(final String appVersion, final String lanternVersion) {
+        versionNum.setText(String.format("%s-%s", appVersion, lanternVersion));
     }
```
Comment (atavism) [inline src/github.com/getlantern/lantern-mobile/app/src/main/java/org/getlantern/lantern/model/UI.java]: After this is called, the text at the bottom of the side-menu will display the app and library version numbers

File: src/github.com/getlantern/lantern-mobile/app/build.gradle [hunk 3]
```diff
@@ -1,20 +1,45 @@
...
+    if ('assembleRelease' in runTasks) {
+        value = 1
+    }
+
+    def code = (versionProps['VERSION_CODE'] ?: "0").toInteger() + value
+    versionProps['VERSION_CODE'] = code.toString()
+    versionProps.store(versionPropsFile.newWriter(), null)
+
+    return code
+}
```
Comment (atavism) [inline src/github.com/getlantern/lantern-mobile/app/build.gradle]: The version code is automatically incremented for every new release build. This gets written to the ``version.properties`` file (which is created and defaults to zero if it doesn't exist already).

**Comment by oxtoacart** [comment]
I get the below error on building android-debug.  I think there are some environment variables and/or gradle properties that I need to configure?  These are worth capturing in the README.

```bash
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/ox.to.a.cart/lantern2/src/github.com/getlantern/lantern-mobile/app/build.gradle' line: 50

* What went wrong:
A problem occurred evaluating project ':app'.
> Could not find property 'KEYSTORE_FILE' on SigningConfig_Decorated{name=release, storeFile=null, storePassword=null, keyAlias=null, keyPassword=null, storeType=null}.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

File: Makefile [hunk 4]
```diff
@@ -569,18 +571,28 @@ $(LANTERN_MOBILE_ANDROID_SDK): $(ANDROID_SDK)
 	mkdir -p $(LANTERN_MOBILE_LIBS) && \
 	cp $(ANDROID_SDK) $(LANTERN_MOBILE_ANDROID_SDK)
 
-$(LANTERN_MOBILE): $(LANTERN_MOBILE_TUN2SOCKS) $(LANTERN_MOBILE_ANDROID_LIB) $(LANTERN_MOBILE_ANDROID_SDK)
+android-dependencies: $(LANTERN_MOBILE_TUN2SOCKS) $(LANTERN_MOBILE_ANDROID_LIB) $(LANTERN_MOBILE_ANDROID_SDK)
```
Comment (oxtoacart) [inline Makefile]: New targets need to be added as prerequisites of .PHONY.

File: Makefile [hunk 5]
```diff
@@ -569,18 +571,28 @@ $(LANTERN_MOBILE_ANDROID_SDK): $(ANDROID_SDK)
...
 	cd $(LANTERN_MOBILE_DIR)/app
-	gradle -b $(LANTERN_MOBILE_DIR)/app/build.gradle \
+	cp $(ANDROID_SDK_ANDROID_LIB) $(LANTERN_MOBILE_LIBS)
+
+android-debug: android-dependencies
+	gradle -PlanternVersion=$(GIT_REVISION) -b $(LANTERN_MOBILE_DIR)/app/build.gradle \
 		clean \
 		assembleDebug
 
-android: $(LANTERN_MOBILE)
```
Comment (oxtoacart) [inline Makefile]: The reason that I used $(LANTERN_MOBILE) is that naming targets after the output file that they create allows Make to be smart about not rebuilding them if the output file already exists.  Setting it up as you have it, we will always rebuild the app whenever we run a dependent target like install.  You can fix this by defining targets for $(LANTERN_MOBILE_ANDROID_DEBUG) and $(LANTERN_MOBILE_ANDROID_RELEASE) and then making android-debug and android-release depend on those.

File: Makefile [hunk 6]
```diff
@@ -569,18 +571,28 @@ $(LANTERN_MOBILE_ANDROID_SDK): $(ANDROID_SDK)
...
+android-release: require-version require-secrets android-dependencies
+	@echo "Generating distribution package for android..."
+	cp $$SECRETS_DIR/android/keystore.release.jks $(LANTERN_MOBILE_DIR)/app
+	gradle -PlanternVersion=$$VERSION -b $(LANTERN_MOBILE_DIR)/app/build.gradle \
+		clean \
+		assembleRelease
+	cp $(LANTERN_MOBILE_ANDROID_RELEASE) lantern-installer.apk
 
-android-debug: $(LANTERN_MOBILE)
+android: android-debug
```
Comment (oxtoacart) [inline Makefile]: The new targets are worth documenting in the README.

File: Makefile (hunk 5 above)
Comment (atavism) [inline Makefile]: Ok, I've updated android-debug and android-release to depend on those new targets.

**Comment by atavism** [comment]
@oxtoacart The properties you need to add to your gradle home ``gradle.properties`` are listed in the README now.

File: README.md [hunk 7]
```diff
@@ -315,6 +318,24 @@ To install on the default device:
 make android-install
 ```
 
+## Release
+
+To create a release build, add the following to your 
+``gradle.properties`` file:
+
+```
+KEYSTORE_PWD=$KEYSTORE_PASSWORD
```
Comment (oxtoacart) [inline README.md]: @atavism FYI, these properties seem to be required even to run a debug build.

Comment (atavism) [inline README.md]: Ok, just fixed that.

**Comment by oxtoacart** [comment]
With a couple of small tweaks that I've committed, this is working great for me.  Merging.

**Pull request merged**